subsys/net/ip/net_shell.c

```diff
@@ -1754,6 +1756,241 @@ static int cmd_net_arp_flush(const struct shell *shell, size_t argc,
 	return 0;
 }
 
+#if defined(CONFIG_NET_CAPTURE)
+static const struct device *capture_dev;
+
+static void get_address_str(const struct sockaddr *addr,
+			    char *str, int str_len)
+{
+	if (IS_ENABLED(CONFIG_NET_IPV6) && addr->sa_family == AF_INET6) {
+		snprintk(str, str_len, "[%s]:%u",
+			 net_sprint_ipv6_addr(&net_sin6(addr)->sin6_addr),
+			 ntohs(net_sin6(addr)->sin6_port));
+
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && addr->sa_family == AF_INET) {
+		snprintk(str, str_len, "%s:%d",
+			 net_sprint_ipv4_addr(&net_sin(addr)->sin_addr),
+			 ntohs(net_sin(addr)->sin_port));
+
+	} else if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) &&
+		   addr->sa_family == AF_PACKET) {
+		snprintk(str, str_len, "AF_PACKET");
+	} else if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN) &&
+		   addr->sa_family == AF_CAN) {
+		snprintk(str, str_len, "AF_CAN");
+	} else if (addr->sa_family == AF_UNSPEC) {
+		snprintk(str, str_len, "AF_UNSPEC");
+	} else {
+		snprintk(str, str_len, "AF_UNK(%d)", addr->sa_family);
+	}
+}
+
+static void capture_cb(struct net_capture_info *info, void *user_data)
+{
+	struct net_shell_user_data *data = user_data;
+	const struct shell *shell = data->shell;
+	int *count = data->user_data;
+	char addr_local[ADDR_LEN + 7];
+	char addr_peer[ADDR_LEN + 7];
+
+	if (*count == 0) {
+		PR("      \t\tCapture  Tunnel\n");
+		PR("Device\t\tiface    iface   Local\t\t\tPeer\n");
+	}
+
+	get_address_str(info->local, addr_local, sizeof(addr_local));
+	get_address_str(info->peer, addr_peer, sizeof(addr_peer));
+
+	PR("%s\t%c        %d      %s\t%s\n", info->capture_dev->name,
+	   info->is_enabled ?
+	   (net_if_get_by_iface(info->capture_iface) + '0') : '-',
+	   net_if_get_by_iface(info->tunnel_iface),
+	   addr_local, addr_peer);
+
+	(*count)++;
+}
+#endif
+
+static int cmd_net_capture(const struct shell *shell, size_t argc,
+			   char *argv[])
+{
+#if defined(CONFIG_NET_CAPTURE)
+	bool ret;
+
+	if (capture_dev == NULL) {
+		PR_INFO("Network packet capture %s\n", "not configured");
+	} else {
+		struct net_shell_user_data user_data;
+		int count = 0;
+
+		ret = net_capture_is_enabled(capture_dev);
+		PR_INFO("Network packet capture %s\n",
+			ret ? "enabled" : "disabled");
+
+		user_data.shell = shell;
+		user_data.user_data = &count;
+
+		net_capture_foreach(capture_cb, &user_data);
+	}
+#else
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	PR_INFO("Set %s to enable %s support.\n",
+		"CONFIG_NET_CAPTURE", "network packet capture");
+#endif
+	return 0;
+}
+
+static int cmd_net_capture_setup(const struct shell *shell, size_t argc,
+				 char *argv[])
+{
+#if defined(CONFIG_NET_CAPTURE)
+	int ret, arg = 1;
+	const char *remote, *local, *peer;
+
+	if (capture_dev != NULL) {
+		PR_INFO("Capture already setup. "
+			"Do \"net capture cleanup\" first.\n");
+		return 0;
+	}
```

Comment:
Can't we just "cleanup" automatically instead of requiring user to call `net capture cleanup` manually?

Reply:
I was thinking this but for now decided to go this route. The API calls have now symmetry, so there is setup() which has cleanup() counterpart, and then enable() and disable().

Reply:
Ah, I actually misunderstood you. Sure, we can have the shell do cleanup automatically.